Limit the number of EDUs in transactions to 100 as expected by receiver

superboum opened the pull request:
### Goal

I have observed this problem on my server:

```
2019-05-04 06:16:57,274 - synapse.federation.sender.transaction_manager - 87 - INFO - federation_transaction_transmission_loop-678370- TX [matrix.livda.fr] {1556878459543} Sending transaction [1556878459543], (PDUs: 0, EDUs: 103)
2019-05-04 06:16:57,338 - synapse.http.matrixfederationclient - 356 - INFO - federation_transaction_transmission_loop-678370- {PUT-O-448448} [matrix.livda.fr] Sending request: PUT matrix://matrix.livda.fr/_matrix/federation/v1/send/1556878459543; timeout 60.000000s
2019-05-04 06:16:57,339 - synapse.http.federation.matrix_federation_agent - 373 - INFO - federation_transaction_transmission_loop-678370- Connecting to matrix.livda.fr:443
[...]
2019-05-04 06:17:08,409 - synapse.http.matrixfederationclient - 388 - INFO - federation_transaction_transmission_loop-678370- {PUT-O-448448} [matrix.livda.fr] Got response headers: 400 Bad Request
2019-05-04 06:17:08,410 - synapse.http.matrixfederationclient - 472 - WARNING - federation_transaction_transmission_loop-678370- {PUT-O-448448} [matrix.livda.fr] Request failed: PUT matrix://matrix.livda.fr/_matrix/federation/v1/send/1556878459543: HttpResponseException("400: b'Bad Request'",)
2019-05-04 06:17:08,411 - synapse.federation.sender.transaction_manager - 123 - INFO - federation_transaction_transmission_loop-678370- TX [matrix.livda.fr] {1556878459543} got 400 response
```

The corresponding issue seems to be *Synapse is not limiting the number of EDUs in a transaction* #3951.
It is a blocker since *Reject large transactions on federation* #4513 as it blocks the federation between two servers.
In my case, the bug was caused by `device_message_edus` that was containing 103 entries.
 
**I don't know synapse internals and just wrote this quick fix to bring back federations. I might have written some wrong/dangerous things.**

### Pull Request Checklist

<!-- Please read CONTRIBUTING.rst before submitting your pull request -->

* [X] Pull request is based on the develop branch
* [X] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.rst#changelog)
* [X] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.rst#sign-off)

richvdh commented on synapse/federation/sender/per_destination_queue.py:
```diff
@@ -217,17 +217,21 @@ def _transaction_transmission_loop(self):
                 pending_edus = []
 
                 pending_edus.extend(self._get_rr_edus(force_flush=False))
+                pending_edus.extend(device_message_edus[:99])
```
I don't think we can safely discard the other `device_message_edus` ?

richvdh commented on synapse/federation/sender/per_destination_queue.py:
```diff
@@ -217,17 +217,21 @@ def _transaction_transmission_loop(self):
-                pending_edus.extend(self._pop_pending_edus(100 - len(pending_edus)))
+                # But we should keep one slot free for presence
+                to_remove = []
+                for key, val in self._pending_edus_keyed.items():
```
I think this would be clearer written as a while loop:

```
while len(pending_edus) < 100:
    _, val = self._pending_edus_keyed.popitem()
    pending_edus.append(val)
```

superboum commented:
Thanks for your reply. I will work on it tonight.
Feel free to directly edit the PR if you need to apply this fix quickly.

superboum commented on synapse/federation/sender/per_destination_queue.py:
```diff
@@ -347,32 +343,33 @@ def _pop_pending_edus(self, limit):
-        contents, stream_id = yield self._store.get_new_device_msgs_for_remote(
-            self._destination, last_device_stream_id, to_device_stream_id
+        last_device_list = self._last_device_list_stream_id
+        now_stream_id, results = yield self._store.get_devices_by_remote(
```
It seems that get_devices_by_remote has a hard coded limit of 20 elements.
So we start with this request to be able to complete with the following one.

richvdh replied:
makes sense. I think it would be helpful to include a comment that says `get_devices_by_remote` will return no more than 20 entries.

superboum commented on synapse/federation/sender/per_destination_queue.py:
```diff
@@ -347,32 +343,33 @@ def _pop_pending_edus(self, limit):
+        last_device_stream_id = self._last_device_stream_id
+        to_device_stream_id = self._store.get_to_device_stream_token()
+        contents, stream_id = yield self._store.get_new_device_msgs_for_remote(
+            self._destination, last_device_stream_id, to_device_stream_id, 98 - len(edus)
```
Here we can set a limit to fill the `edus` list. I chose 98 instead of 100 as `presence` and `rr_edus` could both take one slot after that.

richvdh replied:
I think, rather than hardcode 98 here, it would be clearer to make `_get_new_device_messages` take a `limit` parameter.

richvdh replied:
Also: the default value for the `limit` parameter on `get_new_device_msgs_for_remote` is now redundant. I suggest removing the default value for clarity.

superboum commented:
I have 4 failing unit tests for now (`tox -e py37`) :

```
tests.test_visibility.FilterEventsForServerTestCase.test_large_room
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/home/quentin/Documents/dev/synapse/tests/federation/test_federation_sender.py", line 46, in test_send_receipts
    mock_send_transaction.assert_called_once()
  File "/home/quentin/Documents/dev/synapse/.tox/py37/lib/python3.7/site-packages/mock/mock.py", line 918, in assert_called_once
    raise AssertionError(msg)
builtins.AssertionError: Expected 'send_transaction' to have been called once. Called 0 times.

tests.federation.test_federation_sender.FederationSenderTestCases.test_send_receipts
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/home/quentin/Documents/dev/synapse/tests/federation/test_federation_sender.py", line 81, in test_send_receipts_with_backoff
    mock_send_transaction.assert_called_once()
  File "/home/quentin/Documents/dev/synapse/.tox/py37/lib/python3.7/site-packages/mock/mock.py", line 918, in assert_called_once
    raise AssertionError(msg)
builtins.AssertionError: Expected 'send_transaction' to have been called once. Called 0 times.

tests.federation.test_federation_sender.FederationSenderTestCases.test_send_receipts_with_backoff
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/home/quentin/Documents/dev/synapse/tests/handlers/test_typing.py", line 195, in test_started_typing_remote_send
    try_trailing_slash_on_400=True,
  File "/home/quentin/Documents/dev/synapse/.tox/py37/lib/python3.7/site-packages/mock/mock.py", line 955, in assert_called_once_with
    raise AssertionError(msg)
builtins.AssertionError: Expected 'put_json' to be called once. Called 0 times.

tests.handlers.test_typing.TypingNotificationsTestCase.test_started_typing_remote_send
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/home/quentin/Documents/dev/synapse/tests/handlers/test_typing.py", line 273, in test_stopped_typing
    try_trailing_slash_on_400=True,
  File "/home/quentin/Documents/dev/synapse/.tox/py37/lib/python3.7/site-packages/mock/mock.py", line 955, in assert_called_once_with
    raise AssertionError(msg)
builtins.AssertionError: Expected 'put_json' to be called once. Called 0 times.

tests.handlers.test_typing.TypingNotificationsTestCase.test_stopped_typing
-------------------------------------------------------------------------------
Ran 635 tests in 471.275s

FAILED (skips=1, errors=4, successes=630)
```

I don't know if its due to my modifications. I will check that tomorrow.

richvdh commented on synapse/federation/sender/per_destination_queue.py:
```diff
@@ -248,6 +238,12 @@ def _transaction_transmission_loop(self):
                     )
 
+                pending_edus.extend(device_message_edus)
+                pending_edus.extend(self._pop_pending_edus(100 - len(pending_edus)))
```
this `100` number is starting to appear in a lot of places. Could I suggest creating a constant `MAX_EDUS_PER_TRANSACTION` in this file?

superboum commented:
Thanks a lot for your extensive and precise review. Now tests are passing locally.

 * I added a check for self._pending_edus_keyed directly in the while condition
 * I created a constant as suggested and replaced the hardcoded value by the constant.
 * I added the comment to mention that it will not return more than 20 entries (as I am not very familiar with the code, I am not 100% sure it is the case, so if you have any doubt, we should check before)
 * I added a limit parameter to `_get_new_device_messages` instead of hardcoding 98. I call this function with `MAX_EDUS_PER_TRANSACTION - 2` and added a comment to explain the `- 2` part.
 * I removed the default value for the limit

superboum commented:
To be sure, is the PR blocked on your side (still some review to do, etc.) or on my side (you expect some modifications, like the `assert` one even if >20 is a valid assumption now)?

superboum commented:
To be sure, is the PR blocked on your side (still some review to do, internal process, etc.) or on my side (you expect some modifications, like the `assert` one even if the assumption is valid for now).

superboum commented:
To be sure, is the PR blocked on your side (still some review to do, internal process, etc.) or on my side (you expect some modifications, like the assert one even if the assumption is valid for now)?

richvdh commented:
If you could add the `assert`ion that would be great.
